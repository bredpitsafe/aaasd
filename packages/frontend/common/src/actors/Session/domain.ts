export enum ESessionTypes {
    Auth = 'Auth',
    NotAuth = 'NotAuth',
}

export type TSession = {
    type: ESessionTypes;
};
