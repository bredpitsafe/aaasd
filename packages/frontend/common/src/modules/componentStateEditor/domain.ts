import type { ISO } from '@common/types';

export enum EComponentStateEditorStateSource {
    User = 'User',
    System = 'System',
    Editor = 'Editor',
}

export type TEditorState = {
    source: EComponentStateEditorStateSource;
    revision: ISO;
    selection?: {
        startLine: number;
        startColumn: number;
        endLine: number;
        endColumn: number;
    };
};
