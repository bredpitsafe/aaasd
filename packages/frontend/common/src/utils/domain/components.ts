import { EComponentStatus } from '../../types';
import type { TComponent } from '../../types/domain/component';
import type { TGate } from '../../types/domain/gates';
import type { TRobot } from '../../types/domain/robots';

export function componentIsAllowed(status: EComponentStatus): boolean {
    switch (status) {
        case EComponentStatus.Enabled:
        case EComponentStatus.Alarming:
            return true;
        default:
            return false;
    }
}

export function isRobot(component: TComponent): component is TRobot {
    return isGate(component) === false;
}

export function isGate(component: TComponent): component is TGate {
    return (component as TGate).type !== undefined;
}
