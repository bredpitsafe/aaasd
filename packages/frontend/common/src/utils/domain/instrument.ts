import memoize from 'memoizee';

import type { TInstrument } from '../../types/domain/instrument';

export const getInstrumentByName = memoize(
    (
        instruments: TInstrument[],
        exchangeName: TInstrument['exchange'] | undefined,
        instName: TInstrument['name'] | undefined,
    ) => {
        if (!instName || !exchangeName) {
            return undefined;
        }

        return instruments.find(
            (instr) => instr.exchange === exchangeName && instr.name === instName,
        );
    },
    { max: 100 },
);
