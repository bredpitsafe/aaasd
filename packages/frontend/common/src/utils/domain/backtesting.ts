import { EBacktestingRunStatus } from '../../types/domain/backtestings';

export function backtestingRunCanBeStopped(status: EBacktestingRunStatus) {
    switch (status) {
        case EBacktestingRunStatus.Paused:
        case EBacktestingRunStatus.Running:
        case EBacktestingRunStatus.Initializing:
            return true;
        default:
            return false;
    }
}

export function backtestingRunCanBePaused(status: EBacktestingRunStatus) {
    switch (status) {
        case EBacktestingRunStatus.Running:
            return true;
        default:
            return false;
    }
}

export function backtestingRunCanBeResume(status: EBacktestingRunStatus) {
    switch (status) {
        case EBacktestingRunStatus.Paused:
            return true;
        default:
            return false;
    }
}
