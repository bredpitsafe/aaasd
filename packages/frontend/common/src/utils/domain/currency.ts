// name can be coin/asset name or currency name
import type { ValueOf } from '@common/types';

import type { ECoin, ECurrency } from '../../types/domain/currency';
import { ECoinSymbol, ECurrencySymbol } from '../../types/domain/currency';

export function tryTransformToSymbol<
    N extends string | ValueOf<typeof ECurrency> | ValueOf<typeof ECoin>,
>(name: N): string {
    name = name.toUpperCase() as N;
    return (
        ECurrencySymbol[name as ValueOf<typeof ECurrency>] ??
        ECoinSymbol[name as ValueOf<typeof ECoin>] ??
        name
    );
}
