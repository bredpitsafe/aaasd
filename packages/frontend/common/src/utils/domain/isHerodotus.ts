import type { Nil } from '@common/types';

import type { TRobot } from '../../types/domain/robots';
import { EHerodotusVersion, ERobotClass } from '../../types/domain/robots';

export function isHerodotus(robot: Nil | TRobot): boolean {
    return robot?.kind == ERobotClass.TradingTasks || robot?.kind == ERobotClass.Herodotus;
}

export function getHerodotusVersion(robotKind: string): EHerodotusVersion {
    return robotKind === ERobotClass.Herodotus ? EHerodotusVersion.v2 : EHerodotusVersion.v1;
}
