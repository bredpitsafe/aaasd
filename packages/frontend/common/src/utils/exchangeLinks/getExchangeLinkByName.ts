import type {
    TProviderInstrument,
    TProviderInstrumentDetails,
} from '@backend/bff/src/modules/instruments/schemas/defs.ts';
import { assertNever } from '@common/utils/src/assert.ts';
import { isNil } from 'lodash-es';
import memoize from 'memoizee';

import type { EExchangeName } from '../../types/domain/exchange.ts';
import type { TInstrument, TInstrumentKind } from '../../types/domain/instrument';
import { EInstrumentKindType, EInstrumentStatus } from '../../types/domain/instrument';
import { linkTemplatesFactory } from './linkTemplates';

export const getExchangeLinkByName = memoize(getLink, { max: 100 });

function getLink(instrument: TProviderInstrument & { exchange: string }): URL | undefined;
function getLink(
    instrument: Pick<TInstrument, 'name' | 'status' | 'exchange'> & {
        kind: Pick<TInstrumentKind, 'type'>;
    },
): URL | undefined;
function getLink(
    instrument:
        | (Pick<TInstrument, 'name' | 'status' | 'exchange'> & {
              kind: Pick<TInstrumentKind, 'type'>;
          })
        | (TProviderInstrument & { exchange: string }),
): URL | undefined {
    if (isProviderInstrument(instrument)) {
        const structKind = instrument.details.kind?.type;

        if (isNil(structKind)) {
            return undefined;
        }

        const { name, exchange } = instrument;

        const type = structKindToInstrumentKind(structKind);
        const base = getBaseCurrency(instrument.details);
        const quote = getQuoteCurrency(instrument.details);
        const option = getOptionFromName(name);

        return linkTemplatesFactory[exchange as EExchangeName]?.({
            base,
            quote,
            name,
            option,
            type,
        });
    }

    if (
        instrument.status !== EInstrumentStatus.Trading ||
        linkTemplatesFactory[instrument.exchange] === undefined
    ) {
        return undefined;
    }

    const {
        kind,
        kind: { type },
        name,
        exchange,
    } = instrument;
    const base = 'baseCurrencyName' in kind ? (kind.baseCurrencyName as string) : undefined;
    const quote = 'quoteCurrencyName' in kind ? (kind.quoteCurrencyName as string) : undefined;
    const option = getOptionFromName(name);

    return linkTemplatesFactory[exchange]?.({ base, quote, name, option, type });
}

function isProviderInstrument(
    instrument:
        | (Pick<TInstrument, 'name' | 'status' | 'exchange'> & {
              kind: Pick<TInstrumentKind, 'type'>;
          })
        | TProviderInstrument,
): instrument is TProviderInstrument {
    return 'details' in instrument && 'kind' in instrument.details;
}

function structKindToInstrumentKind(
    structKind: Exclude<TProviderInstrumentDetails['kind'], undefined>['type'],
): EInstrumentKindType {
    switch (structKind) {
        case 'instantSpot':
            return EInstrumentKindType.InstantSpot;
        case 'spotDetails':
            return EInstrumentKindType.Spot;
        case 'futuresDetails':
            return EInstrumentKindType.Futures;
        case 'perpFutures':
            return EInstrumentKindType.PerpFutures;
        case 'option':
            return EInstrumentKindType.Option;
        case 'instrumentSwap':
            return EInstrumentKindType.InstrumentSwap;
        default:
            assertNever(structKind);
    }
}

function getBaseCurrency(details: TProviderInstrumentDetails): string | undefined {
    if (isNil(details.kind)) {
        return undefined;
    }

    const structKind = details.kind.type;

    switch (structKind) {
        case 'instantSpot':
            return details.kind.instantSpot.baseAssetName;

        case 'spotDetails':
            return details.kind.spotDetails.baseAssetName;

        case 'futuresDetails':
        case 'perpFutures':
        case 'option':
        case 'instrumentSwap':
            return undefined;

        default:
            assertNever(structKind);
    }
}

function getQuoteCurrency(details: TProviderInstrumentDetails): string | undefined {
    if (isNil(details.kind)) {
        return undefined;
    }

    const structKind = details.kind.type;

    switch (structKind) {
        case 'instantSpot':
            return details.kind.instantSpot.quoteAssetName;

        case 'spotDetails':
            return details.kind.spotDetails.quoteAssetName;

        case 'futuresDetails':
        case 'perpFutures':
        case 'option':
        case 'instrumentSwap':
            return undefined;

        default:
            assertNever(structKind);
    }
}

function getOptionFromName(name: string): string {
    return name.match(/^(\w+-\w+).*$/)?.[1] ?? '';
}
