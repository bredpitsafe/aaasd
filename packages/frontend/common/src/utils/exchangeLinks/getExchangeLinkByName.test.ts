import { EExchangeName } from '../../types/domain/exchange';
import type { TInstrument } from '../../types/domain/instrument';
import { EInstrumentKindType, EInstrumentStatus } from '../../types/domain/instrument';
import { getExchangeLinkByName } from './getExchangeLinkByName';

describe('getExchangeLinkByName', () => {
    const baseInstrument: Pick<TInstrument, 'name' | 'kind' | 'status' | 'exchange'> = {
        exchange: EExchangeName.BinanceSpot,
        kind: {
            type: EInstrumentKindType.InstantSpot,
            baseCurrency: 123,
            baseCurrencyName: 'BTC',
            quoteCurrency: 456,
            quoteCurrencyName: 'USD',
        },
        name: 'BTCUSD',
        status: EInstrumentStatus.Trading,
    };

    const cases: [EExchangeName, EInstrumentKindType][] = [
        [EExchangeName.BinanceSpot, EInstrumentKindType.InstantSpot],
        [EExchangeName.BinanceSwap, EInstrumentKindType.PerpFutures],
        [EExchangeName.BinanceCoinSwap, EInstrumentKindType.PerpFutures],
        [EExchangeName.KrakenSpot, EInstrumentKindType.InstantSpot],
        [EExchangeName.PoloniexSpot, EInstrumentKindType.InstantSpot],
        [EExchangeName.OkexSpot, EInstrumentKindType.InstantSpot],
        [EExchangeName.OkexSwap, EInstrumentKindType.PerpFutures],
        [EExchangeName.BithumbSpot, EInstrumentKindType.InstantSpot],
        [EExchangeName.HuobiSpot, EInstrumentKindType.InstantSpot],
        [EExchangeName.UpbitSpot, EInstrumentKindType.InstantSpot],
        [EExchangeName.BybitSpot, EInstrumentKindType.InstantSpot],
        [EExchangeName.KucoinSpot, EInstrumentKindType.InstantSpot],
        [EExchangeName.GateioSpot, EInstrumentKindType.InstantSpot],
        [EExchangeName.CoinbaseSpot, EInstrumentKindType.InstantSpot],
    ];

    it.each(cases)(
        'exchange: %p, type: %p',
        (exchange: EExchangeName, type: EInstrumentKindType) => {
            const instrument = {
                ...baseInstrument,
                exchange,
                kind: { ...baseInstrument.kind, type },
            };
            const result = getExchangeLinkByName(instrument);
            expect(result).not.toBeUndefined();
            expect(result?.href).toMatchSnapshot();
        },
    );

    it('should return undefined when status is not `Trading` or `CancelOnly`', () => {
        Object.values(EExchangeName).map((name) => {
            Object.values(EInstrumentStatus).map((status) => {
                if (status === EInstrumentStatus.Trading || EInstrumentStatus.CancelOnly) {
                    return;
                }

                const instrument = {
                    ...baseInstrument,
                    exchange: name,
                    status,
                };
                expect(getExchangeLinkByName(instrument)).toBeUndefined();
            });
        });
    });
});
