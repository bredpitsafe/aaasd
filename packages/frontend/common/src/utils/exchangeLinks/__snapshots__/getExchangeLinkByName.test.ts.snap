// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getExchangeLinkByName exchange: "BinanceCoinSwap", type: "PerpFutures" 1`] = `"https://www.binance.com/en/delivery/BTCUSD"`;

exports[`getExchangeLinkByName exchange: "BinanceSpot", type: "InstantSpot" 1`] = `"https://www.binance.com/en/trade/BTC_USD"`;

exports[`getExchangeLinkByName exchange: "BinanceSwap", type: "PerpFutures" 1`] = `"https://www.binance.com/en/futures/BTCUSD"`;

exports[`getExchangeLinkByName exchange: "BithumbSpot", type: "InstantSpot" 1`] = `"https://en.bithumb.com/trade/order/BTC_USD"`;

exports[`getExchangeLinkByName exchange: "BybitSpot", type: "InstantSpot" 1`] = `"https://www.bybit.com/ru-RU/trade/spot/BTC/USD"`;

exports[`getExchangeLinkByName exchange: "CoinbaseSpot", type: "InstantSpot" 1`] = `"https://exchange.coinbase.com/trade/BTC-USD"`;

exports[`getExchangeLinkByName exchange: "GateioSpot", type: "InstantSpot" 1`] = `"https://www.gate.io/trade/BTC_USD"`;

exports[`getExchangeLinkByName exchange: "HuobiSpot", type: "InstantSpot" 1`] = `"https://www.huobi.com/en-us/exchange/btc_usd"`;

exports[`getExchangeLinkByName exchange: "KrakenSpot", type: "InstantSpot" 1`] = `"https://trade.kraken.com/ru-ru/charts/KRAKEN:BTC-USD"`;

exports[`getExchangeLinkByName exchange: "KucoinSpot", type: "InstantSpot" 1`] = `"https://www.kucoin.com/trade/BTC-USD"`;

exports[`getExchangeLinkByName exchange: "OkexSpot", type: "InstantSpot" 1`] = `"https://www.okx.com/ru/trade-spot/btc-usd"`;

exports[`getExchangeLinkByName exchange: "OkexSwap", type: "PerpFutures" 1`] = `"https://www.okx.com/ru/trade-swap/btc-usd-swap"`;

exports[`getExchangeLinkByName exchange: "PoloniexSpot", type: "InstantSpot" 1`] = `"https://poloniex.com/trade/BTC_USD/?type=spot"`;

exports[`getExchangeLinkByName exchange: "UpbitSpot", type: "InstantSpot" 1`] = `"https://upbit.com/exchange?code=CRIX.UPBIT.USD-BTC"`;
